qiskit/circuit/delay.py

```diff
@@ -54,8 +54,9 @@ def duration(self, duration):
         """Set the duration of this delay."""
         self.params = [duration]
 
-    def to_matrix(self) -> np.ndarray:
+    def __array__(self, dtype=None):
         """Return the identity matrix."""
+        # pylint: disable=unused-argument
         return np.array([[1, 0],
                          [0, 1]], dtype=complex)
```

Comment:
I think we want the default to be complex if `dtype=None`. Otherwise float or int will get used depending on the input matrix.

It makes the code a bit more messy but you could do something something like this for every gate:
```python
if dtype is None:
    dtype = complex
return np.array(..., dtype=dtype) 
```

Reply:
What's wrong with the matrix being float or int since int*complex = complex? Also for ints maybe you reduce floating point errors.

Reply:
Probably nothing